### PR TITLE
migrate coco prototype

### PR DESCRIPTION
**Pull request by pmeier**

### CODE REVIEW

File: torchvision/prototype/datasets/_builtin/imagenet.py [hunk 1]
```diff
@@ -54,7 +54,11 @@ class ImageNetDemux(enum.IntEnum):
 
 @register_dataset(NAME)
 class ImageNet(Dataset2):
-    def __init__(self, root: Union[str, pathlib.Path], *, split: str = "train") -> None:
+    """
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/imagenet.py:57]: Small fix to ImageNet that I found while working on COCO.

File: torchvision/prototype/datasets/utils/_dataset.py [hunk 2]
```diff
@@ -196,7 +196,18 @@ def _verify_str_arg(
...
+    def __init__(
+        self, root: Union[str, pathlib.Path], *, skip_integrity_check: bool = False, dependencies: Collection[str] = ()
+    ) -> None:
+        for dependency in dependencies:
+            try:
+                importlib.import_module(dependency)
+            except ModuleNotFoundError as error:
+                raise ModuleNotFoundError(
+                    f"{type(self).__name__}() depends on the third-party package '{dependency}'. "
+                    f"Please install it, for example with `pip install {dependency}`."
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/utils/_dataset.py:208]: Are we sure all packages are available on PyPI? And are we sure these will be compatible with the user's setup? I'd rather not give explicit instrutions here

File: torchvision/prototype/datasets/utils/_dataset.py [hunk 3]
```diff
@@ -196,7 +196,18 @@ def _verify_str_arg(
...
+                ) from error
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/utils/_dataset.py:209]: from None or from error?

File: torchvision/prototype/datasets/_builtin/coco.py [hunk 4]
```diff
@@ -32,27 +31,49 @@
...
+@register_info(NAME)
+def _info() -> Dict[str, Any]:
+    categories, super_categories = zip(*DatasetInfo.read_categories_file(BUILTIN_DIR / f"{NAME}.categories"))
+    return dict(categories=categories, super_categories=super_categories)
+
+
+@register_dataset(NAME)
+class Coco(Dataset2):
+    """
+    - **homepage**: https://cocodataset.org/
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/coco.py:50]: I would suggest to add the dependencies in the docstring as well

**Review by NicolasHug — APPROVED** [review]
Minor comments but LGTM, thanks @pmeier

File: torchvision/prototype/datasets/utils/_dataset.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/datasets/utils/_dataset.py:208]: > Are we sure all packages are available on PyPI?

I think so yes. I don't think we can depend on packages that are not hosted on PyPI for our public datasets.

> And are we sure these will be compatible with the user's setup? I'd rather not give explicit instrutions here

No, we can't be sure. Given that wheels on PyPI are self-contained, I'm guessing we should be fine. The intention here is to help new users to get started. If we can expect them to handle the installation themselves, I'm ok to remove the instructions.

File: torchvision/prototype/datasets/utils/_dataset.py (hunk 3 above)
Comment (pmeier) [reply torchvision/prototype/datasets/utils/_dataset.py:209]: `from None` is better given that the original error adds no information.

File: torchvision/prototype/datasets/utils/_dataset.py (hunk 2 above)
Comment (NicolasHug) [reply torchvision/prototype/datasets/utils/_dataset.py:208]: Ok let's leave it as is, we can always revisit

**Comment by NicolasHug** [comment]
Tests are green locally, CI is toasted but as discussed offline we'll still merge these migrating PR against the `prototype-datasets-inheritance` branch in order to move faster with this. We'll make sure everything is green on the CI before merging on `main`.

Thanks @pmeier

**Pull request merged**